crates/re_analytics/src/event.rs

```diff
@@ -1,88 +1,232 @@
+
 use std::collections::HashMap;
 
-use time::OffsetDateTime;
-
-/// The "public" API key can be obtained at <https://eu.posthog.com/project/settings#project-api-key>.
-/// Make sure you are logged in to the right organization and have the correct project open.
-/// Unfortunately that stuff is client-side routed, and there's no way to link directly to the right place.
-pub const PUBLIC_POSTHOG_PROJECT_KEY: &str = "phc_sgKidIE4WYYFSJHd8LEYY1UZqASpnfQKeMqlJfSXwqg";
-
-#[derive(Debug, serde::Serialize)]
-#[serde(untagged)]
-pub enum PostHogEvent<'a> {
-    Capture(PostHogCaptureEvent<'a>),
-    Identify(PostHogIdentifyEvent<'a>),
-}
-
-impl<'a> PostHogEvent<'a> {
-    pub fn from_event(analytics_id: &'a str, session_id: &'a str, event: &'a Event) -> Self {
-        let properties = event.props.iter().map(|(name, value)| {
-            (
-                name.as_ref(),
-                match value {
-                    &Property::Integer(v) => v.into(),
-                    &Property::Float(v) => v.into(),
-                    Property::String(v) => v.as_str().into(),
-                    &Property::Bool(v) => v.into(),
-                },
-            )
-        });
-
-        match event.kind {
-            crate::EventKind::Append => Self::Capture(PostHogCaptureEvent {
-                timestamp: event.time_utc,
-                event: event.name.as_ref(),
-                distinct_id: analytics_id,
-                properties: properties
-                    .chain([("session_id", session_id.into())])
-                    .collect(),
-            }),
-            crate::EventKind::Update => Self::Identify(PostHogIdentifyEvent {
-                timestamp: event.time_utc,
-                event: "$identify",
-                distinct_id: analytics_id,
-                properties: [("session_id", session_id.into())].into(),
-                set: properties.collect(),
-            }),
+use re_build_info::BuildInfo;
+
+use crate::{AnalyticsEvent, Event, EventKind, Properties, Property};
+
+impl From<Id> for Property {
+    fn from(val: Id) -> Self {
+        match val {
+            Id::Official(id) => Property::String(id),
+            Id::Hashed(id) => id,
+        }
+    }
+}
+
+impl Event for Identify {
+    const NAME: &'static str = "$identify";
+
+    const KIND: EventKind = EventKind::Update;
+}
+
+impl Properties for Identify {
+    fn serialize(&self, event: &mut AnalyticsEvent) {
+        let Self {
+            build_info,
+            rust_version,
+            llvm_version,
+            python_version,
+            opt_in_metadata,
+        } = self;
+
+        build_info.serialize(event);
+        event.insert_opt("rust_version", rust_version.clone());
+        event.insert_opt("llvm_version", llvm_version.clone());
+        event.insert_opt("python_version", python_version.clone());
+        for (name, value) in opt_in_metadata {
+            event.insert(name.clone(), value.clone());
+        }
```

Comment:
we can avoid a lot of clones if we change `serialize` to consume `self`.
That's how it is used anyway: an event is discarded after it has been serialized.

```suggestion
    fn serialize(self, event: &mut AnalyticsEvent) {
        let Self {
            build_info,
            rust_version,
            llvm_version,
            python_version,
            opt_in_metadata,
        } = self;

        build_info.serialize(event);
        event.insert_opt("rust_version", rust_version);
        event.insert_opt("llvm_version", llvm_version);
        event.insert_opt("python_version", python_version);
        for (name, value) in opt_in_metadata {
            event.insert(name, value);
        }
```